tool/tsh/tsh.go

```diff
@@ -1394,38 +1411,61 @@ func showApps(apps []types.Application, active []tlsca.RouteToApp, verbose bool)
 				app.GetDescription(),
 				app.GetPublicAddr(),
 				app.GetURI(),
-				app.LabelsString(),
+				sortedLabels(app.GetAllLabels()),
 			})
 		}
 		fmt.Println(t.AsBuffer().String())
 	} else {
-		t := asciitable.MakeTable([]string{"Application", "Description", "Public Address", "Labels"})
+		var rows [][]string
 		for _, app := range apps {
-			labelChunks := chunkLabels(app.GetAllLabels(), 2)
-			for i, v := range labelChunks {
-				var name string
-				var addr string
-				if i == 0 {
-					name = app.GetName()
-					addr = app.GetPublicAddr()
-				}
-				for _, a := range active {
-					if name == a.Name {
-						name = fmt.Sprintf("> %v", name)
-					}
+			name := app.GetName()
+			for _, a := range active {
+				if name == a.Name {
+					name = fmt.Sprintf("> %v", name)
 				}
-				t.AddRow([]string{
-					name,
-					app.GetDescription(),
-					addr,
-					strings.Join(v, ", "),
-				})
 			}
+			desc := app.GetDescription()
+			addr := app.GetPublicAddr()
+			labels := sortedLabels(app.GetAllLabels())
+			rows = append(rows, []string{name, desc, addr, labels})
 		}
+		t := makeTableWithTruncatedFinalColumn(
+			[]string{"Application", "Description", "Public Address", "Labels"}, rows)
 		fmt.Println(t.AsBuffer().String())
 	}
 }
 
+func makeTableWithTruncatedFinalColumn(columnOrder []string, rows [][]string) asciitable.Table {
+	t := asciitable.MakeTable([]string{})
+	totalLen := 0
+	for collIndex, colName := range columnOrder[:len(columnOrder)-1] { // final column is handled separately
+		column := asciitable.Column{
+			Title:         colName,
+			MaxCellLength: len(colName),
+		}
+		for _, row := range rows {
+			cellLen := row[collIndex]
+			if len(cellLen) > column.MaxCellLength {
+				column.MaxCellLength = len(cellLen)
+			}
+		}
+		totalLen += column.MaxCellLength + 1
+		t.AddColumn(column)
+	}
+	width, _, err := term.GetSize(int(os.Stdin.Fd()))
+	if err != nil {
+		width = 80
+	}
+	t.AddColumn(asciitable.Column{
+		Title:         columnOrder[len(columnOrder)-1],
+		MaxCellLength: width - totalLen - len("... "),
+	})
```

Comment:
Seems possible for this to result in a negative `MaxCellLength` if the preceeding columns were too wide.  Seems to me that there should be some minimum that `MaxCellLength` cant drop below.